rust-lib/src/lib.rs

```diff
@@ -214,38 +218,50 @@ pub extern "C" fn issue(runtime: &COpaqueStruct, json: *mut c_char) -> CResult {
     _issue(runtime, json).into()
 }
 
-#[derive(Debug, Deserialize)]
-struct TransferArgs {
-    inputs: Vec<OutPoint>,
-    allocate: Vec<Outcoins>,
-    #[serde(with = "serde_with::rust::display_fromstr")]
-    invoice: Invoice,
-    prototype_psbt: String,
-    consignment_file: String,
-    transaction_file: String,
-}
-
-fn _transfer(runtime: &COpaqueStruct, json: *mut c_char) -> Result<(), String> {
+fn _transfer(runtime: &COpaqueStruct, inputs: *mut c_char, allocate: *mut c_char,
+             invoice: *mut c_char, prototype_psbt: *mut c_char, consignment_file: *mut c_char,
+             transaction_file: *mut c_char) -> Result<(), String> {
     let runtime = Runtime::from_opaque(runtime)?;
-    let data: TransferArgs =
-        serde_json::from_str(ptr_to_string(json)?.as_str()).map_err(|e| format!("{:?}", e))?;
-    info!("{:?}", data);
+
+    let inputs: Vec<OutPoint> =
+        serde_json::from_str(ptr_to_string(inputs)?.as_str()).map_err(|e| format!("{:?}", e))?;
+
+    let allocate: Vec<Outcoins> =
+        serde_json::from_str(ptr_to_string(allocate)?.as_str()).map_err(|e| format!("{:?}", e))?;
+
+    let c_invoice = unsafe { CStr::from_ptr(invoice) };
+    let invoice = Invoice::from_str(c_invoice.to_str().unwrap()).unwrap();
```

Comment:
These `unwrap`s is the problem: it causes app to panic on each error. Instead, you have to return Error type with `?` (and you will probably add new error converters for that).

In general, using `unwrap` outside of tests considered very bad practice, making program behavior unpredictable.

Reply:
I've created a `RequestError`, used `amplify_derive` to convert errors to it and removed the `unwrap`s.

now the output of the node.js example looks like this:
```
[2020-10-28T13:02:54Z INFO  rgb] Starting RGB...
[2020-10-28T13:02:54Z INFO  rgb] Config { stash_endpoint: Posix("ipc:/tmp/rgb-node/testnet/stashd.rpc"), contract_endpoints: {Fungible: Posix("/tmp/rgb-node/testnet/fungibled.rpc")}, network: Testnet3, threaded: false, data_dir: "/tmp/rgb-node/" }
d8b8760a32f922310acd086f43c49dbb0aeb2ff25a71f9e83957f3a9fb37073f
InvoiceErr(WrongOutpoint)
```

note: I haven't derived `Clone` on the `RequestError`, because when I try to do that I receive this error:
```
   Compiling rgb_node v0.1.0 (/mnt/dmc/zoe/work/LNP-BP/rgb-node)
error[E0277]: the trait bound `serde_json::Error: Clone` is not satisfied
   --> src/lib.rs:138:14
    |
138 |     SerdeErr(serde_json::Error),
    |              ^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `serde_json::Error`
    |
    = note: required by `clone`
    = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `rgb`
```

I suppose I have to implement `Clone` for `serde_json::Error` but I don't know how to do that